New gas fee escalation mechanism and other improvements

luzzif opened the pull request:
In particular:

- Add `bytea` `answer_tx_hash` column to `active_oracles` table
- Convert address column to `bytea` instead of text to save storage
- Add constraints on the length of both `address` and `answer_tx_hash` columns on the `active_oracles` table
- Avoid showing target in logs (noise) and show the full tree of spans instead of just the current one
- Improve log messages across the board
- Improve task error/join error handling in join sets
- Introduce a gas escalation mechanism to make sure the answer oracle transaction is processed in a timely manner
- While answering an oracle, return early if a previous answering task is still running for it (check if `answer_tx_hash` is there)
- While no answering task is there, start it by submitting a tx and update the answer tx hash to reflect it on the database
- If the answering tx errors out, delete the answer tx hash so that a new answering task can start
- Add tests
- Avoid returning a result from the entrypoint, opt for direct logging of errors + exit(1) so that fatal error logs also have the correct json structure

Commit:
feat: implement escalating tx fees and other things

- add bytea answer_tx_hash column to active_oracles table
- convert address column to bytea instead of text to save storage
- add constraints on the length of both address and answer_tx_hash on the active_oracles table
- avoid showing target in logs (noise) and show the full tree of spans instead of just the current
- improve log messages across the board
- improve task error/join error handling in join sets
- introduce a gas escalation mechanism to make sure the answer oracle transaction is processed in a timely manner
- while answering an oracle, return early if a previous answering task is still running for it (check if answer_tx_hash is there)
- while no answering task is there, start it by submitting a tx and update the answer tx hash to reflect it on the database
- if the answering tx errors out, delete the answer tx hash so that a new answering task can start
- add tests
- avoid returning a result from the entrypoint, opt for direct logging of errors + exit(1) so that fatal error logs also have the correct json structure

Commit:
test: add answer tx hash deletion integration test

Commit:
chore: update log message

guerrap commented on migrations/2023-08-24-074840_create_active_oracles_table/up.sql:
```diff
@@ -1,5 +1,9 @@
+    specification JSONB NOT NULL,
+    answer_tx_hash BYTEA,
+
+    CONSTRAINT address_length CHECK (LENGTH(address) = 20),
```
Shouldn't it be 42 here as the max length?

luzzif replied:
Well 42 bytes is what it takes to represent an address in hex string form, with the 0x prefix, which is the usual "human" representation. In reality, an address is 20 bytes long. The hex string length of 42 comes from the fact that in hex each byte must be represented by 2 hex digits (1 hex digit can represent 16 values max, 0 to 15, which translates to what 4 bits can represent, which is why 2 hex characters are required to represent a byte). That means that the 20 bytes are represented in string form by 40 characters. Then you add the 0x prefix (to make it clear what's following is an hex string) and you end up with 42 chars. Here we're heavily optimizing space as we're storing the bytes directly, without converting them from/to string

guerrap commented on src/scanner/present.rs:
```diff
@@ -240,14 +258,39 @@ async fn answer_active_oracle(
+            Ok(db_connection) => db_connection,
+            Err(error) => {
+                tracing::error!(
+                    "could not get database connection while trying to delete oracle from database - {}",
```
The logged error here should refer to the `answer_tx_hash` update operation, and not the oracle delete, no?

luzzif replied:
It should have been addressed in dab047b95b3723ab37c78f7f4ee2f97179aff9da

Commit:
chore: update log message

Commit:
feat: features and improvements

- update readme
- update defillama oracle abi to make new measurement timestamp function available
- add measurement timestamp colum in the active_oracles table
- in the active_oracles table, make the primary key be a combination of the oracle address and the chain id
- remove get_unix_timestamp utility function
- remove gas escalation feature as currently implemented due to a very reduced support
- replace get_all_for_chain_id with get_all_answerable_for_chain_id, automatically returning all active oracles with measurement timestamp in the past
- handle measurement timestamp as specified on-chain in the oracle instances
- improve logging
- remove timestamp entirely from tvl specification payload
